File: Algorithm/python/class03/RadixSort.py

import sys
sys.path.append("/home/ubuntu/htl/Algorithm/python/include/")
sys.path.append("/home/ubuntu/htl/Algorithm/python/class02/")

from globalFunc import randomArray, selectSort
from Validator import validator

def maxbits(arr):
    maxval = - (2 << 63)
    for i in range(len(arr)):
        maxval = arr[i] if arr[i] > maxval else maxval
    res = 0
    while maxval != 0:
        res += 1
        maxval = maxval // 10
    return res

def getDigit(value, d):
    while d > 1:
        value = value // 10
        d -= 1
    res = value % 10
    return res

def radixSort(arr):
    if arr == None or len(arr) < 2:
        return
    radixSort_process(arr, 0, len(arr) - 1, maxbits(arr))

def radixSort_process(arr, L, R, digit):
    radix = 10
    bucket = [0 for _ in range(R - L + 1)]
    for d in range(1, digit + 1):
        count = [0 for _ in range(radix)]
        for i in range(L, R + 1):
            j = getDigit(arr[i], d)
            count[j] += 1
        for i in range(1, radix):
            count[i] += count[i - 1]
        for i in range(R, L - 1, -1):
            j = getDigit(arr[i], d)
            bucket[count[j] - 1] = arr[i]
            count[j] -= 1
        for i in range(len(arr)):
            arr[i + L] = bucket[i]
        

def main():
    n = 10
    V = 100
    testTimes = 1000
    arr = randomArray(n, V)
    print(arr)
    selectSort(arr)
    print(arr)
    radixSort(arr)
    print(arr)
    validator(selectSort, radixSort, 100, 1000, testTimes)

def test():
    for i in range(10, -1, -1):
        print(i)

main()

File: Algorithm/c/class03/RadixSort.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include "../include/GlobalFuncs.h"


int maxbits(int *arr, int length){

    int maxVal = INT_MIN, res = 0;
    for(int i = 0; i < length; i++){
        maxVal = arr[i] > maxVal ? arr[i] : maxVal;
    }
    while (maxVal != 0)
    {
        res++;
        maxVal /= 10;
    }
    return res;
}

int getDigit(int value, int d){

    int res = 0;
    while (d > 1)
    {
        value /= 10;
        d--;
    }

    res = value % 10;
    return res;
    
}

void radixSort(int *arr, int length){
    if(arr == NULL || length < 2){
        return;
    }
    radixSort_process(arr, 0, length - 1, maxbits(arr, length));
}

void radixSort_process(int *arr, int L, int R, int digit){

    static int radix = 10;          // 基数，这里设置位10
    int i = 0, j = 0;
    int *bucket = (int *)malloc((R - L + 1) * sizeof(arr[0]));
    for(int d = 1; d <= digit; d++){
        int *count = (int *)malloc(radix * sizeof(int));
        for(i = 0; i < radix; i++){
            count[i] = 0;
        }
        for(i = L; i <= R; i++){        // 入桶，这里即用一个数组记录arr元素的第d位值出现的次数
            j = getDigit(arr[i], d);
            count[j]++;
        }
        for(i = 1; i < radix; i++){
            count[i] += count[i - 1];
        }
        for(i = R; i >= L; i--){
            j = getDigit(arr[i], d);
            bucket[count[j] - 1] = arr[i];
            count[j]--;
        }
        for(i = L, j = 0; i <= R; i++, j++){
            arr[i] = bucket[j];
        }

        free(count);
        count = NULL;
    }
    free(bucket);
    bucket = NULL;
}



int main(){


    int N = 10;
    int V = 1000;
    int testTimes = 100000;
    //validator(selectSort, bubbleSort, N, V, testTimes, 1);
    validator(selectSort, radixSort, N, V, testTimes, 1);

}


